marathon_envoy_poc/app.py

```diff
@@ -209,21 +230,79 @@ def endpoints():
         DiscoveryResponse(max_version, cluster_load_assignments, TYPE_EDS))
 
 
-@flask_app.route("/v2/discovery:listeners", methods=["POST"])
-def listeners():
-    # TODO: Without TLS/SNI stuff, this is largely static
-    filter_chains = [
+def http_filter_chains():
+    return [
         FilterChain([
             Filter("envoy.http_connection_manager",
                    HttpConnectionManager("http", "http", own_config_source()))
         ])
     ]
+
+
+def get_certificates():
+    if not hasattr(g, "live_certs"):
+        g.live_certs = {}
+    if not hasattr(g, "certificates"):
+        g.certificates = {}
+
+    vault_client = get_vault()
+    if vault_client is None:
+        flask_app.logger.warn("Unable to fetch certificates: no Vault client.")
+        return g.certificates
+
+    # Get the mapping of domain name to x509 cert hash. This can be used to
+    # check our existing cache of certificates for changes.
+    live_certs = vault_client.get("/live")
+
+    # Regenerate the set of certificates, updating if certs added/changed
+    certificates = {}
+    for domain, cert_id in live_certs.items():
+        if domain in g.live_certs and g.live_certs[domain] == cert_id:
+            # Use the cached cert
+            certificates[domain] = g.certificates[domain]
+        else:
+            # Fetch a new cert
+            # TODO: Validate certificate data in some way
+            certificates[domain] = vault_client.get("/certificates/" + domain)
+        # Removed certs are skipped
+
+    # Update the caches
+    g.live_certs = live_certs
+    g.certificates = certificates
+
+    return g.certificates
+
+
+def https_filter_chains():
+    # NOTE: Filters must be identical across FilterChains
+    filters = [Filter(
+        "envoy.http_connection_manager",
+        HttpConnectionManager("https", "https", own_config_source()))]
```

Comment:
Should we maybe factor out the filters stuff and call `filters("https")` or whatever to guarantee that they're as identical as they need to be?

Reply:
They just need to be identical across a single listener. So the `http` listener can have one set of filters and the `https` listener another...but maybe I should make this clearer in the comment. Also no harm in factoring out.